content/blog/haystack-pipeline-a-deep-dive/index.md

```diff
@@ -0,0 +1,436 @@
+---
+layout: blog-post
+title: Haystack's Pipeline - a deep dive
+description: Building and breaking Haystack's Pipelines to show where they shine and where they fail.
+featured_image: thumbnail.png
+images: ["blog/haystack-pipeline-a-deep-dive/thumbnail.png"]
+alt_image: Several white pipeline graphs on a blue background
+toc: True
+date: 2023-10-15
+last_updated:  2023-10-15
+authors:
+  - Sara Zanzottera
+tags: ["Haystack 2.0"]
+---	
+
+If you've ever looked at Haystack before, you must have come across the concept of Pipeline, one of the most prominent concepts of the framework. However, this abstraction is by no means an obvious choice when it comes to NLP libraries. Why did we adopt this concept, and what does it bring us? 
+
+In this post, I go into all the details of how the Pipeline abstraction works in Haystack now, why it works this way, and its strengths and weaknesses. This deep dive into the current state of the framework is also a premise for the next episode, where I will explain how Haystack 2.0 addresses this version's shortcomings.
+
+If you think you already know how Haystack Pipelines work, give this post a chance: I might manage to change your mind.
+
+## A Bit Of History
+
+Interestingly, in the very first releases of Haystack, Pipelines were not a thing. Version 0.1.0 was released with a simpler object, the [Finder](https://github.com/deepset-ai/haystack/blob/d2c77f307788899eb562d3cb6e42c69b968b9f2a/haystack/__init__.py#L16), that did little more than gluing together a [Retriever](https://docs.haystack.deepset.ai/docs/retriever) and a [Reader](https://docs.haystack.deepset.ai/docs/reader), the two fundamental building blocks of a [semantic search](https://docs.haystack.deepset.ai/docs/glossary#semantic-search) application.
+
+In the next few months, however, the capabilities of language models expanded to enable many more use cases. One hot topic was [hybrid retrieval](https://haystack.deepset.aiblog/hybrid-retrieval): a system composed of two different Retrievers, an optional [Ranker](https://docs.haystack.deepset.ai/docs/ranker), and an optional Reader. This kind of application clearly didn't fit the Finder's design, so in [version 0.6.0](https://github.com/deepset-ai/haystack/releases/tag/v0.6.0) the [Pipeline](https://docs.haystack.deepset.ai/docs/pipelines) object was introduced: a new abstraction that helped users build applications as a graph of components.
```

Comment:
https://haystack.deepset.aiblog/hybrid-retrieval --> the URL needs fixing to `https://haystack.deepset.ai/blog/hybrid-retrieval`